### DISCUSSION TITLE
Access UI / Ingress?

### DISCUSSION THREAD

**Discussion by dvob** [zitadel/zitadel#2575] [Q&A]
I'm trying to setup Zitadel for the first time. To do so I followed the instructions in https://docs.zitadel.ch/docs/guides/installation/crd. 
After the setup I looked for ingress resources in the cluster but I couldn't find any. Am I expected to create them my self or does Ambassador have its own CRD which is equivalent to Ingress?

I then tried to access the UI without ingress as follows but only got back a 404.
```
$ kubectl -n caos-zitadel port-forward service/ui-v1 8080:80
$ curl http://localhost:8080/
404 page not found
```
Under which path/host can I access the UI?

**Comment by fforootd** [comment]
Hi There

 @stebenz or @eliobischof will  provide you with an answer 😁

**Comment by stebenz** [comment]
Yeah, Ambassador has their own CRD which is in this instance equivalent to what you're searching for.
To be precise there are 2 CRDs, Mapping and Host, you can read more about them [here](https://www.getambassador.io/docs/edge-stack/latest/topics/using/intro-mappings/). This CRD are created by the ZITADEL-operator.

The applied CRD also have rewrites which you have to take into account, as for ingress you have to take care to use an ingress which supports GRPC Web, as for example Nginx only supports GRPC Web over a workaround in the configuration.

To have a fully functioning ZITADEL you have to use an Ingress/API Gateway, you can test if ZITADEL is running without but some functions are not available per port-forward.

I hope this answers your questions, if you have others don't hesitate to ask.

**Reply by dvob** [reply]
Hi @stebenz

Thank you for your answer. If I setup Zitadel according to the instructions no hosts(hosts.getambassador.io) and no mappings (mappings.getambassador.io) are created (below I described the steps I did).
I don't necessarily need them because I would prefer to do the setup with Istio, but for that I have to know which ingresses(host/path/rewrites) I have to create. I only took the way with Ambassador because I thought that this the quickest way to setup everything to play around with it.

**steps to reproduce**
```
curl -L -o zitadelctl https://github.com/caos/zitadel/releases/download/v1.48.7/zitadelctl-linux-amd64
curl -L -o orbctl  https://github.com/caos/orbos/releases/download/v5.2.5/orbctl-Linux-x86_64
curl -o database.yaml https://docs.zitadel.ch/assets/files/database-51ebbd568effd89805227c7afc1be4a9.yml
# in database.yaml I ichanged requested memory to 1Gi and storage class to standard
curl -o zitadel.yaml https://docs.zitadel.ch/assets/files/zitadel-9d6f8ed2288d244a079057a4339f02a0.yml
curl -o boom.yaml https://docs.zitadel.ch/assets/files/boom-f6a6a9799873ca962fc3b3e8e35a2a8a.yml
chmod +x zitadelctl orbctl

kind create cluster
./zitadelctl takeoff
kubectl apply -f database.yaml -f zitadel.yaml

cat <<EOF | ./zitadelctl writesecret zitadel.keys.existing --stdin
otpverificationkey_1: $(openssl rand -base64 22)
cookiekey_1: $(openssl rand -base64 22)
domainverificationkey_1: $(openssl rand -base64 22)
idpconfigverificationkey_1: $(openssl rand -base64 22)
oidckey_1: $(openssl rand -base64 22)
userverificationkey_1: $(openssl rand -base64 22)
EOF

./orbctl takeoff
kubectl apply -f boom.yaml
```
Things I noticed:
* The versions in database.yaml and zitadel.yaml are set to  `1.0.0`. I was not sure if I should use the latest version here. In a second try I set them to 1.48.7 but there were still no mappings and hosts.
* After the cockroachdb is ready it takes a long time until the migration pod starts. I'm not sure why this happens?

**Reply by stebenz** [reply]
Hi @dvob 
Sorry that it took me some days to get to this.
I tried to reproduce it, but in my test it created the hosts and mappings without any problems, but it took some time.

With Ambassador you're right, you also have to have the crd-definitions applied in the cluster before the zitadel-operator can apply them, which means the boom.yaml has to be applied beforehand with the "orbctl takeoff" of course.

The point with the version is correct, you can change the version on the fly and the operators will get deployed, but the mappings and hosts also exist in version 1.0.0 so that shouldn't be the problem here.

The db-pods are created by the database-operator but the migration is created by the zitadel-operator as the migrations are application-specific. But before the migrations can be created the certificates for the users have to be created, that's what takes some time, or how much time does it take in your cluster?

Hope I could clarify some things, but for your problem, can you provide me with logs out of the zitadel-operator?
And with a "kubectl api-resources" with at least the ambassador resources to be sure that the CRDs are applied already?

**Reply by machship-mm** [reply]
@stebenz where are the Hosts and Mappings defined in the repo? I am facing the same issue... everything is up and running, but no Host or Mapping showing up in the Ambassador CRDs. I want to try to manually apply the appropriate YAMLs.

**Reply by machship-mm** [reply]
@stebenz 
I'm seeing the following in the zitadel operator logs:
```
ts="2021-11-21T07:47:51Z" err="error while querying: error while querying: error while querying: error while querying: no matches for kind "Host" in version "getambassador.io/v2"" src="/go/src/github.com/caos/zitadel/operator/crtlcrd/zitadel/zitadel.go:32"
```

I can see that it's definitely in the cluster:
```
➜  ~ kubectl get crd | grep getambassador
devportals.getambassador.io                    2021-11-21T07:10:03Z
listeners.getambassador.io                     2021-11-21T07:10:03Z
authservices.getambassador.io                  2021-11-21T07:23:57Z
consulresolvers.getambassador.io               2021-11-21T07:23:58Z
filterpolicies.getambassador.io                2021-11-21T07:23:59Z
filters.getambassador.io                       2021-11-21T07:24:00Z
hosts.getambassador.io                         2021-11-21T07:24:01Z
kubernetesendpointresolvers.getambassador.io   2021-11-21T07:24:02Z
kubernetesserviceresolvers.getambassador.io    2021-11-21T07:24:03Z
logservices.getambassador.io                   2021-11-21T07:24:04Z
mappings.getambassador.io                      2021-11-21T07:24:06Z
modules.getambassador.io                       2021-11-21T07:24:07Z
projectrevisions.getambassador.io              2021-11-21T07:24:07Z
projectcontrollers.getambassador.io            2021-11-21T07:24:07Z
projects.getambassador.io                      2021-11-21T07:24:07Z
ratelimits.getambassador.io                    2021-11-21T07:24:08Z
ratelimitservices.getambassador.io             2021-11-21T07:24:09Z
tcpmappings.getambassador.io                   2021-11-21T07:24:10Z
tlscontexts.getambassador.io                   2021-11-21T07:24:11Z
tracingservices.getambassador.io               2021-11-21T07:24:12Z
```

**Reply by adlerhurst** [reply]
hi @machship-mm 

Thanks for your requests. @eliobischof will have a look at it tomorrow.

**Reply by machship-mm** [reply]
Thanks @adlerhurst. 

Just to provide you guys with some more information, should you want to replicate the issue, these are the steps I'm taking:

To evaluate the system, I have followed the instructions here:
https://docs.zitadel.ch/docs/guides/installation/crd/

So that I can evaluate the process in isolation, I have set up the following:
1. An Ubuntu VM based on [Ubuntu's 18.04 Cloud image](https://cloud-images.ubuntu.com/bionic/current/)  
2. I have given the VM plenty of resources (12 cores, 32GB of RAM)
3. I have installed MicroK8s on that VM using `snap` (to get a single node cluster up and running easily)
4. I have installed the `dns` and `storage` addons for MicroK8s.
5. I changed the `storageClass` in `database.yml` to `microk8s-hostpath` instead of `fast` (note that the storage on the VM is all NVME)
6. I have adjusted the domain name, twilio and email settings in `zitadel.yml` to suit our accounts and platform.
7. I have then run all of the instructions from your page, in the order that they are displayed.

Things to note:
- Nothing else is running on the K8s cluster; it is fresh/clean so that I can ensure nothing else is interfering with your providers.
- Everything seems to start up, the DB migrations take place, etc; however, the logs for the Zitadel operator seem to indicate that it cannot find the CRDs.
- I tried to check the code in the `orbos` repo that the Zitadel Operator uses to interact with K8s to see if I could ascertain whether there are extra filters, etc, but due to it's generic nature I couldn't easily figure that side of it out. I'm considering writing a small service to see if I can replicate what Zitadel Operator is doing and debug the issue.


Let me know if you'd like any more info! As a side note, I got most of the system up and running on my local using docker-compose and from what I can see, you guys have developed a marvelous product!

**Reply by machship-mm** [reply]
Also, FYI, I was using K8s v1.21.

**Reply by eliobischof** [reply]
Thank you for trying ZITADEL and for the detailed information. I'll try to reproduce and debug the issue using the latest operator and ambassador versions.

**Reply by eliobischof** [reply]
@machship-mm I think you have an Ambassador Version 2.x installed. If so, can you please try again using an [Ambassador Version 1.x](https://www.getambassador.io/docs/edge-stack/1.14/tutorials/getting-started/)

I've opened an issue for adding Ambassador v2.x support #2730 

@adlerhurst can you please review this docs improving PR #2731
